design/external-authorization-design.md

```diff
@@ -0,0 +1,670 @@
+
+The result of this is that authorization servers are not supported on HTTP.
+While it is convenient from an implementation perspective, this policy is also consistent with Contour's security-first posture.
+Note that using the TLS fallback certificate (for non-SNI clients) has the same HTTP Connection Manager properties as HTTP, so authentication servers also cannot be configured on virtual hosts that share the fallback certificate.
+
+## Detailed Design
+
+### SupportService Changes
+
+The `SupportService` CRD has API version `projectcontour.io/v1alpha1`.
+
+There are a number of benefits to creating a CRD to represent a supporting service:
+
+- The CRD directly generates an Envoy [Cluster][4] that Contour can use for any purpose.
+  This means that with a single new API, Contour can add support for authorization, rate limiting and logging support services.
+- A CRD gives Contour a way to communicate the operational status of the support service, which was a [desired][9] [goal][7].
+- A CRD allows the team operating the support service and the team operating Contour to collaborate more loosely.
+
+```Go
+// SupportProtocolVersion is the version of the GRPC protocol used
+// to request support services.
+type SupportProtocolVersion string
+
+// SupportProtocolVersion2 requests the "v2" support protocol version.
+const SupportProtocolVersion2 SupportProtocolVersion = "v2"
+
+type SupportService struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   SupportServiceSpec   `json:"spec,omitempty"`
+	Status SupportServiceStatus `json:"status,omitempty"`
+}
+
+type SupportServiceStatus struct {
+	Conditions []Condition `json:"conditions"`
+}
+
+type SupportServiceSpec struct {
+	// +required
+	Service contourv1.Service `json:"service"`
+
+	// The load balancing policy for sending requests to the service.
+	// +optional
+	LoadBalancerPolicy *contourv1.LoadBalancerPolicy `json:"loadBalancerPolicy,omitempty"`
+
+	// This field sets the version of the GRPC protocol that Envoy uses to
+	// send requests to the support service. Since Contour always uses the
+	// v2 Envoy API, this is currently fixed at "v2". However, other
+	// protocol options will be available in future.
+	//
+	// +kubebuilder:validation:Enum=v2
+	ProtocolVersion SupportProtocolVersion `json:"protocolVersion,omitempty"`
+}
+```
+
+The `Conditions` field in `SupportServicesStatus` follows the standardized Contour structure (proposed in #2642).
+This type should be shared with other Contour API status types.
+
+This document does not yet define how Contour monitors the support service.
+However, the presence of the `Conditions` field allows Contour to expose the status of the support service fairly directly.
+At minimum, Contour should observe the underlying Kubernetes Service and expose whether there are sufficient healthy Endpoints.
```

Comment:
how is "sufficient healthy Endpoints" defined/determined?

Reply:
Probably greater than 0. I expect this to evolve over time.